src/libtexture/texturesys.cpp

```diff
@@ -2147,7 +2147,7 @@ TextureSystemImpl::sample_bilinear(
         if (onetile && all(stvalid)) {
             // Shortcut if all the texels we need are on the same tile
             id.xy(sttex[S0] - tile_st[S0], sttex[T0] - tile_st[T0]);
-            bool ok = find_tile(id, thread_info);
+            bool ok = find_tile(id, thread_info, sample == 0);
```

Comment:
Same question here actually -- if the `find_tile` calls would always hit the same tile, the call would be outside the loop, no?

So why is it ok to only flag the first tile we found as used?

Is there maybe a different way to structure all of this to recognize when the `id` is changing from iteration to iteration?

Reply:
The thread remembers which tile it saw last... but "last" may be from a previous texture call quite a while ago. So we want to poke the "used" flag on THIS call, but we need not do it multiple times. So this means that the first (of maybe several) probes we do on this call will be sure to set "used", even if it's the same tile we last accessed on a prior call.

If this texture call ends up accessing more than one tile for this call, then that later probe (even if not probe number 0) will be one where the tile it needs isn't the "last" tile, so it too will poke the new tile's used flag.

Reply:
Let me repeat this in my own words just to make sure I have it right...

The code in `master` calls `use()` on every access. This patch only calls `use()` for the first sample - and if subsequent calls hit the same tile, they will skip the `use()` call (because it knows it can't be the same one).

If we cross over into another tile, `find_tile_main_cache` will do the call to `use()` on its own. And any subsequent probes into that tile would by definition not need to twiddle `use()` again.

I'm having a hard time reasoning about the call to `use()` after the swap from `lasttile`, but I think it doesn't need the same logic, right? By definition it can only happen "once" for a sequence of probes that hit the same tile.

Reply:
yeah, I think you have it